### PR TITLE
Do not show IDE CTAs to users who have these products installed

### PR DESCRIPTION
**Pull request by philipp-spiess**
Fixes https://github.com/sourcegraph/sourcegraph/issues/30843

We do not want to show IDE CTAs for users that we know are already users of our IDE extensions.

In discussions prior to creating this PR we [ruled out the option of using VSCE specific logs](https://github.com/sourcegraph/sourcegraph/pull/31876), most importantly because we noticed that only ~25% of our VSCE users properly have their Sourcegraph account linked (and it's unclear if this is even possible for JetBrains?).

Instead, this PR takes a different approach. We're using the fact that all links coming from an IDE extension and opening in the browser have their UTM tags added. With this we can "tag" the currently logged in user's extension usage ([or use localStorage if no one is logged in](https://docs.sourcegraph.com/dev/background-information/web/temporary_settings)).

To avoid stale data, we require a link to be opened in the last 30 days in order to count the user as an active extension user.

## Test plan

We (@vdavid and I) manually tested this by changing the conditions of the CTA banners to always render and then using the following three example links and verify they correctly store the extension usage in a user's temporary settings:

- `https://sourcegraph.test:3443/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph-jetbrains.git&branch=main&file=src%2Fmain%2Fjava%2FOpenRevisionAction.java&editor=JetBrains&version=v1.2.2&start_row=68&start_col=26&end_row=68&end_col=26&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.3.2`
- `https://sourcegraph.test:3443/-/editor?remote_url=git@github.com:sourcegraph/sourcegraph.git&branch=ps/detect-ide-extensions&file=client/web/src/tracking/util.ts&editor=VSCode&version=2.0.9&start_row=13&start_col=22&end_row=13&end_col=22&utm_campaign=vscode-extension&utm_medium=direct_traffic&utm_source=vscode-extension&utm_content=vsce-commands`
- `https://sourcegraph.test:3443/sign-up?editor=vscode&utm_medium=VSCIDE&utm_source=sidebar&utm_campaign=vsce-sign-up&utm_content=sign-up`

Additionally, we've added a test for the component:

```
α sourcegraph (ps/detect-ide-extensions)✗ yarn jest IdeExtensionTracker
yarn run v1.22.17
$ /Users/philipp/dev/sourcegraph/node_modules/.bin/jest IdeExtensionTracker
 PASS   web  client/web/src/IdeExtensionTracker.test.tsx
  IdeExtensionTracker
    ✓ Detects the proper extension for "https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph-jetbrains.git&branch=main&file=src%2Fmain%2Fjava%2FOpenRevisionAction.java&editor=JetBrains&version=v1.2.2&start_row=68&start_col=26&end_row=68&end_col=26&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.3.2" (13 ms)
    ✓ Detects the proper extension for "https://sourcegraph.com/-/editor?remote_url=git@github.com:sourcegraph/sourcegraph.git&branch=ps/detect-ide-extensions&file=client/web/src/tracking/util.ts&editor=VSCode&version=2.0.9&start_row=13&start_col=22&end_row=13&end_col=22&utm_campaign=vscode-extension&utm_medium=direct_traffic&utm_source=vscode-extension&utm_content=vsce-commands" (2 ms)
    ✓ Detects the proper extension for "https://sourcegraph.com/sign-up?editor=vscode&utm_medium=VSCIDE&utm_source=sidebar&utm_campaign=vsce-sign-up&utm_content=sign-up" (2 ms)
    ✓ Detects the proper extension for "https://sourcegraph.com/?something=different" (2 ms)

Test Suites: 1 passed, 1 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        1.009 s
Ran all test suites matching /IdeExtensionTracker/i.
✨  Done in 2.00s.
```

### CODE REVIEW

File: client/web/src/IdeExtensionTracker.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import { useTemporarySetting } from '@sourcegraph/shared/src/settings/temporary/useTemporarySetting'
+
+/**
+ * This component uses UTM parameters to detect incoming traffic from our IDE extensions (VS Code
+ * and JetBrains) and updates a temporary setting whenever these are found.
+ */
+export const IdeExtensionTracker: React.FunctionComponent = () => {
+    const [, setLastVSCodeDetection] = useTemporarySetting(VSCODE_SETTING, 0)
+    const [, setLastJetBrainsDetection] = useTemporarySetting(JETBRAINS_SETTING, 0)
```
Comment (limitedmage) [inline client/web/src/IdeExtensionTracker.tsx:15]: I wonder how plausible it would be to set these temporary settings inside the IDE extensions themselves. instead of trying to read events from the log on the webapp. I don't think we have temporary settings available in the extensions (yet) but to me it doesn't feel like too much extra work. Maybe can be done as a follow-up to this?

File: client/web/src/IdeExtensionTracker.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    // We only want to run the below effect once. Since the setter function changes over time, we
+    // capture a copy in a ref to avoid passing it into the dependency array of the effect.
+    const setLastVSCodeDetectionReference = useRef(setLastVSCodeDetection)
+    const setLastJetBrainsDetectionReference = useRef(setLastJetBrainsDetection)
```
Comment (limitedmage) [inline client/web/src/IdeExtensionTracker.tsx:20]: Is the reason this is required a bug in temporary settings?

File: client/web/src/IdeExtensionTracker.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            setLastJetBrainsDetectionReference.current(Date.now())
+        } else if (utmMedium === 'VSCIDE' || utmSource?.toLowerCase().startsWith('vscode')) {
+            setLastVSCodeDetectionReference.current(Date.now())
+        }
+    }, [])
+
+    return null
+}
+
+export const useIsUsingIdeIntegration = (): undefined | boolean => {
```
Comment (limitedmage) [inline client/web/src/IdeExtensionTracker.tsx:38]: Maybe this can be renamed to be more clear that this is use is recent? eg. `useIsUsingIdeIntegrationThisMonth` (though that's a bit long...)

File: client/web/src/IdeExtensionTracker.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import React, { useEffect, useRef, useState } from 'react'
+
+const VSCODE_SETTING = 'integrations.vscode.lastDetectionTimestamp'
+const JETBRAINS_SETTING = 'integrations.jetbrains.lastDetectionTimestamp'
```
Comment (limitedmage) [inline client/web/src/IdeExtensionTracker.tsx:4]: These shouldn't be needed as `useTemporarySetting` should restrict the strings via TypeScript typechecking `TemporarySettingsSchema`

File: client/web/src/repo/actions/InstallIntegrationsAlert.tsx [hunk 5]
```diff
@@ -42,6 +43,7 @@ export const InstallIntegrationsAlert: React.FunctionComponent<InstallIntegratio
         IDE_CTA_CADENCE_SHIFT
     )
     const isBrowserExtensionInstalled = useObservable<boolean>(browserExtensionInstalled)
+    const isUsingIdeIntegration = useIsUsingIdeIntegration()
     const [hoverCount] = useLocalStorage<number>(HOVER_COUNT_KEY, 0)
     const [hasDismissedBrowserExtensionAlert, setHasDismissedBrowserExtensionAlert] = useLocalStorage<boolean>(
         HAS_DISMISSED_BROWSER_EXTENSION_ALERT_KEY,
```
Comment (limitedmage) [inline client/web/src/repo/actions/InstallIntegrationsAlert.tsx:49]: Consider changing these dismissals to use temporary settings instead of localstorage. For existing users, we have the functionality to [migrate from localstorage to temporary settings](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/settings/temporary/migrateLocalStorageToTemporarySettings.ts?L40%3A23-40%3A61=).

File: client/web/src/IdeExtensionTracker.test.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+const settingsClient = createMockClient(
+    { contents: JSON.stringify({}) },
+    gql`
+        query {
+            temporarySettings {
+                contents
+            }
+        }
+    `
+)
```
Comment (limitedmage) [inline client/web/src/IdeExtensionTracker.test.tsx:23]: Use `MockTemporarySettings` instead of `TemporarySettingsContext.Provider` to avoid having to do this yourself.

File: client/web/src/IdeExtensionTracker.test.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        ],
+        ['https://sourcegraph.com/?something=different', null],
+    ]
+    test.each(cases)('Detects the proper extension for %p', async (url, expectedResult) => {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        delete window.location
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        window.location = new URL(url)
```
Comment (limitedmage) [inline client/web/src/IdeExtensionTracker.test.tsx:51]: Why is this required?

**Review by limitedmage — APPROVED** [review]
Looks great! Left some comments and suggestions.

File: client/web/src/IdeExtensionTracker.tsx (hunk 1 above)
Comment (philipp-spiess) [reply client/web/src/IdeExtensionTracker.tsx:15]: Yeah we're thinking of doing this as a second step like I mentioned here: https://github.com/sourcegraph/sourcegraph/pull/31876#issuecomment-1058187363

The reason why we have prioritized this approach is that currently only ~25% of our VSCE users are signed-in. That's another issue we want to improve though 🙂

File: client/web/src/IdeExtensionTracker.tsx (hunk 2 above)
Comment (philipp-spiess) [reply client/web/src/IdeExtensionTracker.tsx:20]: I thought it was reasonable when I looked at the implementation quickly but maybe it's not? I'll take another look, ty!

File: client/web/src/IdeExtensionTracker.test.tsx (hunk 7 above)
Comment (philipp-spiess) [reply client/web/src/IdeExtensionTracker.test.tsx:51]: The React component looks at the `location` to get the current query params and [this](https://sourcegraph.com/search?q=context:global+r:%5Egithub%5C.com/sourcegraph/sourcegraph%24++window.location+%3D+&patternType=literal) is how you mock that 😐

Comment (limitedmage) [reply client/web/src/IdeExtensionTracker.test.tsx:51]: Can we mock it with `MemoryRouter`, `createMemoryHistory`, `createLocation`?

Comment (limitedmage) [reply client/web/src/IdeExtensionTracker.test.tsx:51]: For this, you should use `useLocation` instead of `window.location` directly.

File: client/shared/src/settings/temporary/TemporarySettings.ts [hunk 8]
```diff
@@ -20,6 +20,8 @@ export interface TemporarySettingsSchema {
     'signup.finishedWelcomeFlow': boolean
     'codemonitor.info.visible': boolean
     'homepage.userInvites.tab': number
+    'integrations.vscode.lastDetectionTimestamp': number
+    'integrations.jetbrains.lastDetectionTimestamp': number
```
Comment (rrhyne) [inline client/shared/src/settings/temporary/TemporarySettings.ts:24]: Is there a reason we aren't doing this for Atom and Sublime text as well? Do we have very few users of these extensions or are we not planning to support them in the future?

**Review by rrhyne — APPROVED** [review]
Looks great and I think this approach is much better!!

File: client/shared/src/settings/temporary/TemporarySettings.ts (hunk 8 above)
Comment (philipp-spiess) [reply client/shared/src/settings/temporary/TemporarySettings.ts:24]: The main reason is probably that we haven't thought about it and also aren't using Atom and Sublime. Do we know overall usage of these extensions? Is it worth adding it?

**Review by vdavid — APPROVED** [review]
Looks great! :) (Finally found the time to look at this PR :)

File: client/web/src/search/results/StreamingSearchResults.tsx [hunk 9]
```diff
@@ -134,7 +136,12 @@ function useCtaAlert(
             return 'browser'
         }
 
-        if (!hasDismissedIDEExtensionAlert && displayIDEExtensionCTABasedOnCadence) {
+        if (
+            isUsingIdeIntegration === false &&
+            displayIDEExtensionCTABasedOnCadence &&
+            !hasDismissedIDEExtensionAlert &&
+            true
```
Comment (philipp-spiess) [inline client/web/src/search/results/StreamingSearchResults.tsx:143]: rm

File: client/web/src/IdeExtensionTracker.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+export const IdeExtensionTracker: React.FunctionComponent = () => {
+    const location = useLocation()
+
+    const [, setLastVSCodeDetection] = useTemporarySetting('integrations.vscode.lastDetectionTimestamp', 0)
+    const [, setLastJetBrainsDetection] = useTemporarySetting('integrations.jetbrains.lastDetectionTimestamp', 0)
+
+    // We only want to capture the IDE UTM parameters on the first page load. In order to avoid
+    // rerunning the effect below whenever location change, we instead capture a reference that we
+    // never update.
+    const locationReference = useRef(location)
```
Comment (erzhtor) [inline client/web/src/IdeExtensionTracker.tsx:21]: Also, we can use `location` directly in `useEffect`, but not include it in dependencies.

File: client/web/src/IdeExtensionTracker.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        const parameters = new URLSearchParams(locationReference.current.search)
+        const utmProductName = parameters.get('utm_product_name')
+        const utmMedium = parameters.get('utm_medium')
+        const utmSource = parameters.get('utm_source')
+
+        if (utmProductName === 'IntelliJ IDEA') {
+            setLastJetBrainsDetection(Date.now())
+        } else if (utmMedium === 'VSCIDE' || utmSource?.toLowerCase().startsWith('vscode')) {
+            setLastVSCodeDetection(Date.now())
+        }
```
Comment (erzhtor) [inline client/web/src/IdeExtensionTracker.tsx:33]: Just to clarify, do already use these UTM parameters on VSCE/Intellij Idea extensions or should we add them as a follow-up?

**Review by erzhtor — APPROVED** [review]
Looks good, great job! 👍

File: client/web/src/IdeExtensionTracker.tsx (hunk 11 above)
Comment (vdavid) [reply client/web/src/IdeExtensionTracker.tsx:33]: We already use them.
